pymatgen/analysis/xas/spectrum.py

```diff
@@ -2,57 +2,178 @@
 __copyright__ = "Copyright 2012, The Materials Project"
-__version__ = "2.0"
-__maintainer__ = "Chen Zheng"
-__email__ = "chz022@ucsd.edu"
-__date__ = "Aug 9, 2017"
+__version__ = "3.0"
+__maintainer__ = "Yiming Chen"
+__email__ = "chz022@ucsd.edu, yic111@ucsd.edu"
+__date__ = "Jan 26, 2020"
 
 
-class XANES(Spectrum):
+class XAS(Spectrum):
     """
-    Basic XANES object.
+    Basic XAS object.
 
     Args:
         x: A sequence of x-ray energies in eV
         y: A sequence of mu(E)
-        structure (Structure): Structure associated with the XANES.
-        absorption_specie (Specie): Specie associated with the XANES.
-        edge (str):
+        structure (Structure): Structure associated with the spectrum
+        absorption_element (Element): Element associated with the spectrum
+        edge (str): Absorption edge associated with the spectrum
+        spectrum_type (str): 'XANES' or 'EXAFS'
+
 
     .. attribute: x
         The sequence of energies
 
     .. attribute: y
         The sequence of mu(E)
 
-    .. attribute: absorption_specie
-        The absorption_species of the spectrum
+    .. attribute: absorption_element
+        The absorption_element of the spectrum
 
     .. attribute: edge
-        The edge of XANES spectrum
+        The edge of the spectrum
+
+    .. attribute: spectrum_type
+        XANES or EXAFS spectrum
+
     """
     XLABEL = 'Energy'
     YLABEL = 'Intensity'
 
-    def __init__(self, x, y, structure, absorption_specie, edge):
-        super().__init__(x, y, structure, absorption_specie, edge)
+    def __init__(self, x, y, structure, absorption_element, edge="K",
+                 spectrum_type="XANES"):
+        """
+        Initializes a spectrum object.
+        """
+
+        super(XAS, self).__init__(x, y, structure, absorption_element,
+                                  edge)
         self.structure = structure
-        self.absorption_specie = absorption_specie
+        self.absorption_element = absorption_element
         self.edge = edge
-        self.e0 = self.x[np.argmax(np.gradient(self.y) / np.gradient(self.x))]
+        self.spectrum_type = spectrum_type
+        self.e0 = self.x[np.argmax(np.gradient(self.y) /
+                                   np.gradient(self.x))]
+        # Wavenumber, k is calculated based on equation
+        #             k^2=2*m/(hbar)^2*(E-E0)
+        self.k = [np.sqrt((i-self.e0) / 3.8537) if i > self.e0 else
+                  -np.sqrt((self.e0-i) / 3.8537) for i in self.x]
 
     def __str__(self):
-        return "%s %s Edge for %s: %s" % (
-            self.absorption_specie, self.edge,
+        return "%s %s Edge %s for %s: %s" % (
+            self.absorption_element, self.edge, self.spectrum_type,
             self.structure.composition.reduced_formula,
-            super().__str__()
+            super(XAS, self).__str__()
         )
+
+    @classmethod
+    def stitch_xanes_exafs(cls, xanes, exafs, num_samples=500):
```

Comment:
This should not be a class method. It should simply be XAS.stitch(self, other, num_samples=500). I would prefer we don't have two different stitch methods. You can have a parameter called mode that defines whether it is the XANES/EXAFS or the L23.

Reply:
You should be able to detect what to do based on edge, absorption element, and type automatically correct? If the user provides a set of spectra that are not valid, then throw an error.

Reply:
Technically yes, but I can see different requirements. E.g., some people may want to stitch in different ways, etc. Right now, there is no difference in XANES and EXAFS in terms of class. So other than the spectrum_type, there is nothing to distinguish them. I would hesitate to try to determine the stitching automatically based on the spectrum_type and let the user be explicit, with a sensible default.

Reply:
I'm pretty sure we're the only ones that have the data to perform stitching, so it might be useful to have `stitch` be automatic and then have `stitch_` variants for the different methods. `stitch` should also do some sanity checking that the user isn't trying to stitch a K-edge XANES with a M-edge EXAFS for instance.

Reply:
I agree that we might be the only ones to have data for stitching, but whole XAS class is not only about stitching. I believe a more general usage is to simply obtain those already implemented attributes/methods while automatic stitching could delay such processes and provide data more than enough. 
Thus, I would keep `stitch` a separate method inside XAS class with modes for XANES+EXAFS and L23 as well as more sanity check.